cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -118,25 +119,38 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
 
 // Sync synchronizes client's endpoints with the known endpoints from the etcd membership.
 func (c *Client) Sync() error {
-	cli, err := clientv3.New(clientv3.Config{
-		Endpoints:   c.Endpoints,
-		DialTimeout: 20 * time.Second,
-		TLS:         c.TLS,
-	})
-	if err != nil {
-		return err
+	var cli *clientv3.Client
+	var err error
+
+	for _, endpoint := range c.Endpoints {
+		cli, err = clientv3.New(clientv3.Config{
+			Endpoints:   []string{endpoint},
+			DialTimeout: 20 * time.Second,
+			TLS:         c.TLS,
+		})
+		if err != nil {
+			return err
+		}
+
+		// check client/endpoint health before synching against it
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		_, err = cli.Get(ctx, "health")
+		cancel()
+		if err == nil || err == rpctypes.ErrPermissionDenied {
+			break
```

Comment:
i think `cli *clientv3.Client` needs to be inside the loop and that you need to better manage Close on e.g. break.

Reply:
@neolit123 Maybe you're seeing something that I don't here, but I disagree.
We need the `cli *clientv3.Client` after the loop, so it needs to be defined before that block. It is reassigned on every iteration to do the healthcheck on that node, and if the healthcheck **passes** we break the loop and keep that good client. If it **fails** we continue and `Close()` the client.